File: bridge/src/platforms/discord.js
'use strict';

const { Client, GatewayIntentBits } = require('discord.js');
const { execute } = require('../executor');
const { sanitize } = require('../utils/sanitize');
const { chunk } = require('../utils/chunker');

const DISCORD_MAX_LEN = 1990;

/**
 * Start the Discord platform adapter.
 * @param {object} config - Discord configuration block from bridge.config.json.
 */
async function start(config) {
  if (!config || !config.enabled || !config.token) {
    console.log('[discord] Disabled or no token — skipping.');
    return null;
  }

  const {
    token,
    command_prefix = '!claude ',
    allowed_channel_ids = [],
    allowed_user_ids = [],
  } = config;

  const client = new Client({
    intents: [
      GatewayIntentBits.Guilds,
      GatewayIntentBits.GuildMessages,
      GatewayIntentBits.MessageContent,
    ],
  });

  client.on('ready', () => {
    console.log(`[discord] Logged in as ${client.user.tag}`);
  });

  client.on('messageCreate', async (message) => {
    // Ignore bots
    if (message.author.bot) return;

    // Channel check
    if (allowed_channel_ids.length > 0 && !allowed_channel_ids.includes(message.channel.id)) {
      return;
    }

    // User check
    if (allowed_user_ids.length > 0 && !allowed_user_ids.includes(message.author.id)) {
      return;
    }

    const botMention = `<@${client.user.id}>`;
    const botMentionNick = `<@!${client.user.id}>`;
    const isMentioned =
      message.content.startsWith(botMention) ||
      message.content.startsWith(botMentionNick);
    const isPrefixed = message.content.startsWith(command_prefix);

    if (!isPrefixed && !isMentioned) return;

    // Extract prompt
    let rawPrompt;
    if (isPrefixed) {
      rawPrompt = message.content.slice(command_prefix.length);
    } else {
      // Strip mention
      rawPrompt = message.content
        .replace(botMention, '')
        .replace(botMentionNick, '')
        .trim();
    }

    const { safe: prompt } = sanitize(rawPrompt);
    if (!prompt.trim()) return;

    let processingMsg;
    try {
      processingMsg = await message.reply('⏳ 處理中...');
    } catch (err) {
      console.error('[discord] Failed to send processing message:', err);
      return;
    }

    const result = await execute(prompt, null);

    const output = result.success
      ? result.output || '(no output)'
      : `Error: execution failed.\n${result.output}`;

    const chunks = chunk(output, DISCORD_MAX_LEN);

    // Edit the first placeholder message with the first chunk
    try {
      await processingMsg.edit(chunks[0]);
    } catch (err) {
      console.error('[discord] Failed to edit processing message:', err);
    }

    // Send remaining chunks as follow-up replies
    for (let i = 1; i < chunks.length; i++) {
      try {
        await message.reply(chunks[i]);
      } catch (err) {
        console.error(`[discord] Failed to send chunk ${i}:`, err);
      }
    }
  });

  client.on('error', (err) => {
    console.error('[discord] Client error:', err);
  });

  await client.login(token);
  return client;
}

module.exports = { start };

File: bridge-go/discord.go
package main

import (
	"context"
	"fmt"
	"log/slog"
	"regexp"
	"slices"
	"strings"
	"sync"

	"github.com/bwmarrin/discordgo"
)

const discordChunkRunes = 1990

var mentionRe = regexp.MustCompile(`<@!?[0-9]+>`)

type Discord struct {
	cfg  DiscordConfig
	exec ExecutorConfig
	log  *slog.Logger
	wg   sync.WaitGroup
}

func NewDiscord(cfg DiscordConfig, exec ExecutorConfig) *Discord {
	return &Discord{cfg: cfg, exec: exec, log: slog.With("platform", "discord")}
}

func (d *Discord) Name() string { return "discord" }

func (d *Discord) Run(ctx context.Context) error {
	session, err := discordgo.New("Bot " + d.cfg.Token)
	if err != nil {
		return fmt.Errorf("init session: %w", err)
	}

	session.Identify.Intents = discordgo.IntentsGuilds |
		discordgo.IntentsGuildMessages |
		discordgo.IntentsMessageContent
	session.ShouldReconnectOnError = true

	session.AddHandler(func(s *discordgo.Session, m *discordgo.MessageCreate) {
		d.wg.Add(1)
		defer d.wg.Done()
		defer func() {
			if r := recover(); r != nil {
				d.log.Error("handler panic", "recover", r)
			}
		}()
		d.handleMessage(ctx, s, m)
	})
	session.AddHandler(func(s *discordgo.Session, r *discordgo.Ready) {
		d.log.Info("bot ready", "username", r.User.Username)
	})

	if err := session.Open(); err != nil {
		return fmt.Errorf("open session: %w", err)
	}
	// Wait for in-flight handlers before closing the session.
	defer d.wg.Wait()
	defer session.Close()

	d.log.Info("bot started")
	<-ctx.Done()
	return nil
}

func (d *Discord) handleMessage(ctx context.Context, s *discordgo.Session, m *discordgo.MessageCreate) {
	// Ignore bots
	if m.Author.Bot {
		return
	}

	// State.User is nil before the Ready event fires
	if s.State == nil || s.State.User == nil {
		return
	}

	// Channel whitelist
	if len(d.cfg.AllowedChannelIDs) > 0 && !slices.Contains(d.cfg.AllowedChannelIDs, m.ChannelID) {
		return
	}
	// User whitelist
	if len(d.cfg.AllowedUserIDs) > 0 && !slices.Contains(d.cfg.AllowedUserIDs, m.Author.ID) {
		return
	}

	content := m.Content
	botID := s.State.User.ID
	prefixMention1 := "<@" + botID + ">"
	prefixMention2 := "<@!" + botID + ">"

	var rawPrompt string
	switch {
	case strings.HasPrefix(content, d.cfg.CommandPrefix):
		rawPrompt = strings.TrimPrefix(content, d.cfg.CommandPrefix)
	case strings.HasPrefix(content, prefixMention1) || strings.HasPrefix(content, prefixMention2):
		rawPrompt = mentionRe.ReplaceAllString(content, "")
		rawPrompt = strings.TrimSpace(rawPrompt)
	default:
		return
	}

	rawPrompt = strings.TrimSpace(rawPrompt)
	if rawPrompt == "" {
		return
	}

	prompt, truncated := sanitizePrompt(rawPrompt)
	if prompt == "" {
		return
	}

	// Notify user if prompt was truncated before processing.
	if truncated {
		if _, err := s.ChannelMessageSend(m.ChannelID, "⚠️ 您的訊息過長，已截斷至 10,000 字元。"); err != nil {
			d.log.Error("send truncation notice failed", "err", err)
		}
	}

	// Send placeholder
	placeholder, err := s.ChannelMessageSend(m.ChannelID, "⏳ 處理中...")
	if err != nil {
		d.log.Error("send placeholder failed", "err", err)
		return
	}

	result := Execute(ctx, d.exec, prompt)
	output := result.FormatForDisplay()
	chunks := Chunk(output, discordChunkRunes)

	// Edit placeholder with first chunk
	edit := &discordgo.MessageEdit{
		Channel: m.ChannelID,
		ID:      placeholder.ID,
		Content: &chunks[0],
	}
	if _, err := s.ChannelMessageEditComplex(edit); err != nil {
		d.log.Error("edit placeholder failed", "err", err)
		// Fallback: send as a new message
		if _, err2 := s.ChannelMessageSend(m.ChannelID, chunks[0]); err2 != nil {
			d.log.Error("fallback send failed", "err", err2)
			return
		}
	}

	// Send remaining chunks; stop and notify on failure
	for i, chunk := range chunks[1:] {
		ref := m.Reference()
		msg := &discordgo.MessageSend{Content: chunk, Reference: ref}
		if _, err := s.ChannelMessageSendComplex(m.ChannelID, msg); err != nil {
			d.log.Error("send chunk failed", "chunk", i+1, "err", err)
			notice := &discordgo.MessageSend{Content: "⚠️ 回覆傳送不完整，部分內容遺失。", Reference: ref}
			_, _ = s.ChannelMessageSendComplex(m.ChannelID, notice)
			return
		}
	}
}
